C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29847 C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests**
* #29846 Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h
* #29845 C++ tensor indexing: add Slice / TensorIndex

This PR adds three things:
- `Tensor::idx()` and `Tensor::idx_put_()` APIs, which mirror Python API `[]`-indexing getter and setter. There are two reasons for naming it `idx()` instead of `index()` (and `idx_put_()` instead of `index_put_()`):
  - We currently already have `index()` and `index_put_()` APIs in the Tensor class, which accept `ArrayRef<Tensor>` as the indices and only do advanced indexing. If we were to add overloads to `index()` and `index_put_()` to accept `ArrayRef<TensorIndex>` and `std::initializer_list<TensorIndex>` as indices, we will be silently changing the behavior if the user originally has code that does e.g. `tensor.index({tensor1, tensor2})`, since now this goes to the `std::initializer_list<TensorIndex>` path (which has different behavior) instead of the original `ArrayRef<Tensor>` path.
  - `idx()` and `idx_put_()` are easier to type than `index()` and `index_put_()`. As tensor indexing is a very commonly used API, we should make the API as easy to use as possible.
- Mechanisms for parsing `ArrayRef<TensorIndex>` indices and performing indexing operations, mirroring the functions in `torch/csrc/autograd/python_variable_indexing.cpp`. We will be able to merge parts of the C++ and Python indexing code paths in the future, after making sure there is no perf regression when doing so.
- Simple tests to demonstrate that the `Tensor::idx()` and `Tensor::idx_put_()` APIs work. I will add more tests after the first few PRs are reviewed.

Differential Revision: [D18694890](https://our.internmc.facebook.com/intern/diff/D18694890)

Commit:
C++ tensor advanced indexing: add idx() and idx_put_() overloads, and simple indexing tests

Commit:
Update on "C++ tensor advanced indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

kostmo commented:
## CircleCI build failures summary
As of commit ef14330:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=ef14330bfb99d1d04a12bb46051e515a5ea2b8e7):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_xenial_py3_clang5_asan_test | Test | `[  FAILED  ] TensorIndexingTest.TensorIndex` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

Commit:
Update on "C++ tensor indexing: add Tensor::idx() and Tensor::idx_put_(), and simple indexing tests"

C++ tensor indexing: add idx() and idx_put_() overloads, and simple indexing tests

gh-metadata: pytorch pytorch 29847 gh/yf225/74/head

yf225 commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/30425.
